### PR TITLE
459 audit faults etc

### PR DESCRIPTION
**Pull request by hwindsor**
## Pull Request Overview

- Audit logs REST APIs implemented 
- Many more index columns added 
- Audit logs have a durationn

---
Jon & Colin for the code changes
Ben for the API

**Database Updates **

Audit Logs have a duration
All created columns are indexable on all houston tables
Many more columns in AuditLog indexable and limits removed from message length

**REST API Updates **

A new example POST API is defined with its expected parameters, if needed to be specified for a thorough review

```
[GET] /api/v1/audit_log (querystring set to module_name='Sighting')
Equates to 'Tell me the sightings with the most recent audit updates'
[
    {'module_name': 'Sighting', 'item_guid': '7178c998-3ae0-47bf-8522-e41bb2d994af'}, 
    {'module_name': 'Sighting', 'item_guid': '0bc2b3c4-d553-4936-91a3-76f66d7ef256'}, 
    {'module_name': 'Sighting', 'item_guid': '70be095d-b561-4b24-a596-5fa9a61adb8d'}, 
    {'module_name': 'Sighting', 'item_guid': '4589627c-f0e0-448d-9213-8804774eb83b'}, 
    {'module_name': 'Sighting', 'item_guid': '299b999d-8fef-4194-a749-69709ea79e7b'}, 
    {'module_name': 'Sighting', 'item_guid': '153574de-d1c3-4519-b6db-d97183745274'}, 
    {'module_name': 'Sighting', 'item_guid': '2052a343-6339-4e20-9aff-c63b60246a29'}, 
    {'module_name': 'Sighting', 'item_guid': 'a5d805d2-f7c9-430d-8124-867802224403'}, 
    {'module_name': 'Sighting', 'item_guid': '0829cb48-f008-4c97-a539-998a5a81cff2'}, 
    {'module_name': 'Sighting', 'item_guid': '0be88133-3596-4649-8acc-3bf04ec5983e'}, 
    {'module_name': 'Sighting', 'item_guid': 'f72a47d1-503c-461e-97e8-2e11dc37fb97'}, 
    {'module_name': 'Sighting', 'item_guid': 'd2a18868-e9cc-4e32-a4b7-4e99685a5d86'}, 
    {'module_name': 'Sighting', 'item_guid': '60570c77-75dc-4ea7-9a97-21fc6bd280f5'}, 
    {'module_name': 'Sighting', 'item_guid': '34204c1d-a5eb-400b-b420-a3dbf8c5855b'}, 
    {'module_name': 'Sighting', 'item_guid': '77f5ba6d-0108-4210-beed-341effcf8eed'}, 
    {'module_name': 'Sighting', 'item_guid': 'fa4643a5-5125-4868-9190-45e393e5085b'}, 
    {'module_name': 'Sighting', 'item_guid': '5ea28013-27fd-4c69-8c0b-ab6f34382c04'}, 
    {'module_name': 'Sighting', 'item_guid': 'd856c0cc-82fc-41dc-a38c-6a2df56db7d7'}, 
    {'module_name': 'Sighting', 'item_guid': '42a720fa-bcb8-437b-976e-65de2cde1ec9'}
]

[GET] /api/v1/audit_log/{sighting UUID}
Equates to 'Tell me the last audit entries for this sighting'

[
    {
	'message': 'Failed: AssetGroupMetadata uploadType field missing from request 400',
	'audit_type': 'Front End Fault',
	'module_name': None,
	'item_guid': None,
	'duration': None,
	'guid': '8e0a4d6d-4f87-4b99-b64b-17024763f413',
	'user_email': 'researcher1@localhost',
	'created': '2021-09-03T11:13:50.232787+00:00'
    },
    {
	'message':
	'Failed: AssetGroupMetadata assetReferences incorrect type in Sighting 1 400',
	'audit_type': 'Front End Fault',
	'module_name': None,
	'item_guid': None,
	'duration': None,
	'guid': '9477da11-923a-4fff-aa69-be3bf6ed1223',
	'user_email': 'researcher1@localhost',
	'created': '2021-09-03T11:13:50.212859+00:00'},
    {
	'message':
	'Failed: AssetGroupMetadata Encounter 1.1 needs to be a dict 400',
	'audit_type': 'Front End Fault',
	'module_name': None,
	'item_guid': None,
	'duration': None,
	'guid': '6ac5a949-dc0a-448e-a90c-86768d789fc1',
	'user_email': 'researcher1@localhost',
	'created': '2021-09-03T11:13:50.192929+00:00'
    },
    {
	'message':
	'Failed: AssetGroupMetadata encounters field missing from Sighting 1 400',
	'audit_type': 'Front End Fault',
	'module_name': None,
	'item_guid': None,
	'duration': None,
	'guid': 'a3462607-9ee9-47b8-b97f-481a2eb1e061',
	'user_email': 'researcher1@localhost',
	'created': '2021-09-03T11:13:50.158948+00:00'
    },
    {
	'message':
	'Failed: AssetGroupMetadata startTime field missing from Sighting 1 400',
	'audit_type': 'Front End Fault',
	'module_name': None,
	'item_guid': None,
	'duration': None,
	'guid': '11235ed5-a6f3-4bc6-bb67-ee0340855797',
	'user_email': 'researcher1@localhost',
	'created': '2021-09-03T11:13:50.137557+00:00'
    },
    {
	'message':
	'Failed: AssetGroupMetadata Sighting 1 needs to be a dict 400',
	'audit_type': 'Front End Fault',
	'module_name': None,
	'item_guid': None,
	'duration': None,
	'guid': 'a3b83cb8-46d3-475a-90bd-dcc632ac5790',
	'user_email': 'researcher1@localhost',
	'created': '2021-09-03T11:13:50.093836+00:00'
    },
    {
	'message':
	'Failed: AssetGroupMetadata sightings in request must have at least one entry 400',
	'audit_type': 'Front End Fault',
	'module_name': None,
	'item_guid': None,
	'duration': None,
	'guid': '759a0390-bb7c-480e-a256-bcae040b6edc',
	'user_email': 'researcher1@localhost',
	'created': '2021-09-03T11:13:50.073511+00:00'
    },
    {
	'message':
	'Failed: AssetGroupMetadata sightings field missing from request 400',
	'audit_type': 'Front End Fault',
	'module_name': None,
	'item_guid': None,
	'duration': None,
	'guid': '104af1d0-67e8-4503-bfaf-974ad987e143',
	'user_email': 'researcher1@localhost',
	'created': '2021-09-03T11:13:50.051753+00:00'},
    {
	'message':
	'Failed: AssetGroupMetadata speciesDetectionModel field missing from request 400',
	'audit_type': 'Front End Fault',
	'module_name': None,
	'item_guid': None,
	'duration': None,
	'guid': 'af919e40-fa07-470e-9dcb-871e749269be',
	'user_email': 'researcher1@localhost',
	'created': '2021-09-03T11:13:50.031056+00:00'
    },
    {
	'message':
	'Failed: AssetGroupMetadata uploadType field missing from request 400',
	'audit_type': 'Front End Fault',
	'module_name': None,
	'item_guid': None,
	'duration': None,
	'guid': 'ed672a7d-922d-4407-8f84-33b1d2f1a872',
	'user_email': 'researcher1@localhost',
	'created': '2021-09-03T11:13:49.600159+00:00'},
    {
	'message':
	'Failed: AssetGroupMetadata assetReferences incorrect type in Sighting 1 400',
	'audit_type': 'Front End Fault',
	'module_name': None,
	'item_guid': None,
	'duration': None,
	'guid': 'd5cc17fa-49cd-407c-b1f4-47c017207f86',
	'user_email': 'researcher1@localhost',
	'created': '2021-09-03T11:13:49.579908+00:00'
    },
    {
	'message':
	'Failed: AssetGroupMetadata Encounter 1.1 needs to be a dict 400',
	'audit_type': 'Front End Fault',
	'module_name': None,
	'item_guid': None,
	'duration': None,
	'guid': '37c7633d-ccda-4dac-9c98-168ef650b169',
	'user_email': 'researcher1@localhost',
	'created': '2021-09-03T11:13:49.560243+00:00'
    },
    {
	'message':
	'Failed: AssetGroupMetadata encounters field missing from Sighting 1 400',
	'audit_type': 'Front End Fault',
	'module_name': None,
	'item_guid': None,
	'duration': None,
	'guid': '8a74aa2e-c89c-4d30-a8a8-92ef8d99dbe9',
	'user_email': 'researcher1@localhost',
	'created': '2021-09-03T11:13:49.537617+00:00'
    },
    {
	'message':
	'Failed: AssetGroupMetadata startTime field missing from Sighting 1 400',
	'audit_type': 'Front End Fault',
	'module_name': None,
	'item_guid': None,
	'duration': None,
	'guid': '6f379df7-32ec-4e58-9814-3a0464b578ff',
	'user_email': 'researcher1@localhost',
	'created': '2021-09-03T11:13:49.506285+00:00'
    },
    {
	'message':
	'Failed: AssetGroupMetadata Sighting 1 needs to be a dict 400',
	'audit_type': 'Front End Fault',
	'module_name': None,
	'item_guid': None,
	'duration': None,
	'guid': 'ebdcbe1b-a396-4a08-8ebb-b5dd47ddfeca',
	'user_email': 'researcher1@localhost',
	'created': '2021-09-03T11:13:49.481121+00:00'
    },
    {
	'message':
	'Failed: AssetGroupMetadata sightings in request must have at least one entry 400',
	'audit_type': 'Front End Fault',
	'module_name': None,
	'item_guid': None,
	'duration': None,
	'guid': '129d847a-9127-465b-b85c-7ac7d5f24fd0',
	'user_email': 'researcher1@localhost',
	'created': '2021-09-03T11:13:49.444962+00:00'
    },
    {
	'message':
	'Failed: AssetGroupMetadata sightings field missing from request 400',
	'audit_type': 'Front End Fault',
	'module_name': None,
	'item_guid': None,
	'duration': None,
	'guid': 'cb65bcd1-e92a-495b-ac7b-85332f6c460c',
	'user_email': 'researcher1@localhost',
	'created': '2021-09-03T11:13:49.410180+00:00'
    },
    {
	'message':
	'Failed: AssetGroupMetadata speciesDetectionModel field missing from request 400',
	'audit_type': 'Front End Fault',
	'module_name': None,
	'item_guid': None,
	'duration': None,
	'guid': '2c302ef2-68b5-4412-8da5-bd7913e89f28',
	'user_email': 'researcher1@localhost',
	'created': '2021-09-03T11:13:49.388656+00:00'
    },
    {
	'message':
	'Failed: AssetGroupSighting b7c91478-7675-49b7-a828-470adf62bc0e is not detecting 400',
	'audit_type': 'Front End Fault',
	'module_name': None,
	'item_guid': None,
	'duration': None,
	'guid': '20242b4b-8265-4440-a1ac-19ef2dba2dab',
	'user_email': 'internal@localhost',
	'created': '2021-09-03T11:07:57.207322+00:00'
    },
    {
	'message':
	'Failed: AssetGroupMetadata Encounter 1.1 individual c8104aca-d7f1-4c81-ac16-582ee4512baf not found 400',
	'audit_type': 'Front End Fault',
	'module_name': None,
	'item_guid': None,
	'duration': None,
	'guid': '0508c421-2e76-4b9e-ab4a-0d69cf8cb010',
	'user_email': 'anonymous user',
	'created': '2021-09-03T11:07:50.678722+00:00'
    }
]


```

### CODE REVIEW

File: app/modules/audit_logs/parameters.py [hunk 1]
```diff
@@ -8,7 +8,13 @@
 from app.extensions.api.parameters import PaginationParameters
 
 
-class GetCollaborationParameters(PaginationParameters):
+class GetAuditLogParameters(PaginationParameters):
```
Comment (hwindsor) [inline app/modules/audit_logs/parameters.py:11]: See, you guys failed the test I set to see if you spotted this :rofl:

**Review by brmscheiner — APPROVED** [review]
API lgtm

**Review by colinwkingen — APPROVED** [review]
This all LGTM. What is the plan for getting all (or N..) logs for a specific UUID and class?

Approved, but I'll ask you to take a look at the test failure here before merging.

File: app/modules/annotations/resources.py [hunk 2]
```diff
@@ -68,6 +68,9 @@ def post(self, args):
         """
         from app.modules.assets.models import Asset
         from app.modules.encounters.models import Encounter
+        from app.extensions.elapsed_time import ElapsedTime
+
+        timer = ElapsedTime()
```
[the remark this answers is not in the document]
Comment (hwindsor) [reply app/modules/annotations/resources.py:73]: I'd rather not. We don't need precision down to the ms here, we need a "oh boy that looks wrong" big red flag. It doesn't make much difference if it take 2 minutes and 2 seconds or 2 minutes and 3 seconds. the fact that it takes over 2 minutes is the big problem.

File: app/modules/audit_logs/parameters.py (hunk 1 above)
Comment (hwindsor) [reply app/modules/audit_logs/parameters.py:11]: Well either I set a booby trap to see if you were paying attention or I missed it myself and spotted it second time around. 
Go on, have a guess which is most likely :rofl:

File: migrations/versions/f0258425e851_audit_and_timestamp_indecies.py [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+    with op.batch_alter_table('asset_group', schema=None) as batch_op:
+        batch_op.create_index(batch_op.f('ix_asset_group_created'), ['created'], unique=False)
+        batch_op.create_index(batch_op.f('ix_asset_group_updated'), ['updated'], unique=False)
+
+    with op.batch_alter_table('asset_group_sighting', schema=None) as batch_op:
+        batch_op.create_index(batch_op.f('ix_asset_group_sighting_created'), ['created'], unique=False)
+        batch_op.create_index(batch_op.f('ix_asset_group_sighting_updated'), ['updated'], unique=False)
+
+    with op.batch_alter_table('audit_log', schema=None) as batch_op:
+        batch_op.add_column(sa.Column('duration', sa.Float(), nullable=True))
```
Comment (hwindsor) [inline migrations/versions/f0258425e851_audit_and_timestamp_indecies.py:49]: There is something missing here. The max length has been dropped from the user_email and length in the model definition but the change is not reflected here. 
This is what is causing the test failures on github I think. 
houston=# \d audit_log
                          Table "public.audit_log"
   Column    |            Type             | Collation | Nullable | Default 
-------------+-----------------------------+-----------+----------+---------
 created     | timestamp without time zone |           | not null | 
 updated     | timestamp without time zone |           | not null | 
 guid        | uuid                        |           | not null | 
 module_name | character varying(50)       |           |          | 
 item_guid   | uuid                        |           |          | 
 user_email  | character varying(120)      |           | not null | 
 message     | character varying(240)      |           |          | 
 audit_type  | character varying           |           | not null | 
 duration    | double precision            |           |          | 
This is on my local machine where the tests work perfectly. No I don't understand either.

File: app/modules/annotations/resources.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (colinwkingen) [reply app/modules/annotations/resources.py:73]: Yeah, I didn't mean for precision really, just to make sure it was working as expected. Not an issue I'll press though.

**Comment by colinwkingen** [comment]
Looks like the addition of compare_type=True in run_migrations_offline() didn't help. We might be at @bluemellophone 's plan to drop the column and make a new one.

File: migrations/env.py [hunk 4]
```diff
@@ -53,6 +53,7 @@ def run_migrations_offline():
     context.configure(
         url=url,
         render_as_batch=True,
+        compare_type=True,
```
Comment (karenc) [inline migrations/env.py:56]: I checked and `compare_type` is actually already set to be true here:

https://github.com/WildMeOrg/houston/blob/ea1a1ca63b9b23d77dc131e3e79e9fab1a78f2f3/app/extensions/flask_sqlalchemy/__init__.py#L82-L84

So I don't think this line actually does anything

(I checked the value of `self._user_compare_type` in `alembic/runtime/migration.py`)

File: app/modules/audit_logs/models.py [hunk 5]
```diff
@@ -21,16 +18,20 @@ class AuditLog(db.Model, Timestamp):
...
     )  # pylint: disable=invalid-name
-    module_name = db.Column(db.String(length=50), nullable=True)
+    module_name = db.Column(db.String(length=50), index=True, nullable=True)
 
     # Item and user guids intentionally not backrefs to other models as we want to maintain the log even after
     # the item has been removed
-    item_guid = db.Column(db.GUID, nullable=True)
-    user_email = db.Column(db.String(length=120), nullable=False)
+    item_guid = db.Column(db.GUID, index=True, nullable=True)
+    user_email = db.Column(db.String(), nullable=False)
```
Comment (karenc) [inline app/modules/audit_logs/models.py:26]: If you want to remove the length you can change from `db.String` to `db.Text`.  (I checked, a migration is generated)  The database type will change from `VARCHAR` to `TEXT` and I think `VARCHAR` requires a maximum length.

**Review by karenc — COMMENTED** [review]
It appears you're making a lot of changes (one I can see is adding indexes to Timestamp), could you try to separate them into different commits next time?

File: app/modules/audit_logs/models.py (hunk 5 above)
Comment (karenc) [reply app/modules/audit_logs/models.py:26]: Oh actually it looks like that's not true:

```
houston=# \d audit_log    
                          Table "public.audit_log"
   Column    |            Type             | Collation | Nullable | Default
-------------+-----------------------------+-----------+----------+---------
 created     | timestamp without time zone |           | not null |
 updated     | timestamp without time zone |           | not null |
 guid        | uuid                        |           | not null |
 module_name | character varying(50)       |           |          |
 item_guid   | uuid                        |           |          |
 user_email  | character varying(120)      |           | not null |
 message     | character varying(240)      |           |          |
 audit_type  | character varying           |           | not null |
Indexes:
    "pk_audit_log" PRIMARY KEY, btree (guid)

```

`audit_type` is `VARCHAR` but doesn't have a length ... :thinking:

File: migrations/env.py (hunk 4 above)
Comment (hwindsor) [reply migrations/env.py:56]: Confirmed. I reran the migration with tthis change and the migration file is identical.

File: app/modules/audit_logs/models.py (hunk 5 above)
Comment (hwindsor) [reply app/modules/audit_logs/models.py:26]: Gone with the drop and recreate the column idea instead. This is not a problem in this case as there will be no existing audit logs anywhere in the DBs that anyone cares about but in future we do need to know how to do this stuff.

**Comment by hwindsor** [comment]
> It appears you're making a lot of changes (one I can see is adding indexes to Timestamp), could you try to separate them into different commits next time?
I will try to remember to do that in future. I had issues with one of my recent PRs where the commits all got duplicated in a rebase somehow so I squashed them all.

**Review by naknomum — APPROVED** [review]
looks great

**Pull request merged**